docs(mempool/lanes): Add technical design section to ADR

hvanz opened the pull request:
Closes #3138 

[Rendered](https://github.com/cometbft/cometbft/blob/hvanz/mempool-qos-technical-design/docs/references/architecture/adr-113-mempool-lanes.md)

---

#### PR checklist

- [X] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -96,10 +97,11 @@ Before stating this property, we need to make some definitions.
-transactions in all classes.
+a *default class*, which is a special class always present in any set of classes. This is analogous
+to the _native VLAN_ for untagged traffic in an 802.1Q network. Because no transaction can belong to
+two or more classes, transaction classes form disjoint sets, that is, the intersection between
```
```suggestion
two or more classes, transaction classes form disjoint sets, that is, the intersection between any two
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
-      will be gating (e.g., `TxsAvailable`)
+This sections describes the architectural changes needed to implement an MVP of lanes in the
+mempool. The following is a summary of the key design decisions:
+- [[1](#lanes-definition)] The list of lanes and their corresponding priorities
```
```suggestion
- [[Lanes definition](#lanes-definition)] The list of lanes and their corresponding priorities
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+mempool. The following is a summary of the key design decisions:
+- [[1](#lanes-definition)] The list of lanes and their corresponding priorities
+  will be hardcoded in the application logic.
+- [[2](#internal-data-structures)] There will be one concurrent list (CList) data structure per
```
```suggestion
- [[Internal data structures](#internal-data-structures)] There will be one concurrent list (CList) data structure per
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+  will be hardcoded in the application logic.
+- [[2](#internal-data-structures)] There will be one concurrent list (CList) data structure per
+  lane.
+- [[3](#configuration)] All lanes will share the same mempool configuration.
```
```suggestion
- [[Configuration](#configuration)] All lanes will share the same mempool configuration.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+- [[2](#internal-data-structures)] There will be one concurrent list (CList) data structure per
+  lane.
+- [[3](#configuration)] All lanes will share the same mempool configuration.
+- [[4](#adding-transactions-to-the-mempool)] When validating a transaction via CheckTx, the
```
```suggestion
- [[Adding transactions to the mempool](#adding-transactions-to-the-mempool)] When validating a transaction via CheckTx, the
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+- [[3](#configuration)] All lanes will share the same mempool configuration.
+- [[4](#adding-transactions-to-the-mempool)] When validating a transaction via CheckTx, the
+  application will optionally return a lane for the transaction.
+- [[5](#reaping-transactions-for-creating-blocks)] Transactions will be reaped from higher-priority
```
```suggestion
- [[Reaping transactions for creating blocks](#reaping-transactions-for-creating-blocks)] Transactions will be reaped from higher-priority
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+  application will optionally return a lane for the transaction.
+- [[5](#reaping-transactions-for-creating-blocks)] Transactions will be reaped from higher-priority
+  lanes first, preserving the FIFO ordering.
+- [[6](#transaction-dissemination)] We will continue to use the current P2P channel for
```
```suggestion
- [[Transaction dissemination](#transaction-dissemination)] We will continue to use the current P2P channel for
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+- [[4](#adding-transactions-to-the-mempool)] When validating a transaction via CheckTx, the
+  application will optionally return a lane for the transaction.
+- [[5](#reaping-transactions-for-creating-blocks)] Transactions will be reaped from higher-priority
+  lanes first, preserving the FIFO ordering.
```
```suggestion
  lanes first, preserving intra-lane FIFO ordering.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
-    * BUT, some things in the
-      _[modularity laundry list](https://www.notion.so/informalsystems/8a887f27e40b45dead689f2d1762b778?v=30a6fd81673b483fb16f4b7a9fc311fb)_
-      will be gating (e.g., `TxsAvailable`)
+This sections describes the architectural changes needed to implement an MVP of lanes in the
```
```suggestion
This section describes the architectural changes needed to implement an MVP of lanes in the
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
````diff
@@ -294,254 +296,285 @@ TODO
+```protobuf
+message InfoResponse {
+  ...
+  repeated uint32 lanes = 6;
````
`lane_priorities` maybe?

hvanz replied:
Changed to `lane_priorities` and `default_lane_priority`.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+The mempool does not care about the names, only about the priorities. That is why the lane
+information returned by the application only contains priorities.
+
+The lowest priority a lane may have is 1. The value 0 is reserved for two cases: when there are no
+priorities and for `CheckTx` responses of invalid transactions.
```
```suggestion
The lowest priority a lane may have is 1. The value 0 is reserved for two cases: when the application does not classify the transaction (i.e. no priority returned) and for `CheckTx` responses of invalid transactions.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+also needs to define which of the lanes it defines is the default lane. 
+
+To obtain the lane information from the application, we need to extend the ABCI `Info` response to
+include the following fields. These fields need to be filled by the application only in case it
```
```suggestion
include the following fields.
Remember CometBFT calls `Info` when the node is starting up (or recovering) as part of CometBFT handshake with the application.
These fields need to be filled by the application only in case it
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+
+On receiving the information from the app, CometBFT will validate that:
+- `lanes` has no duplicates (values in `lanes` don't need to be sorted),
+- `default_value` is in `lanes` (the default lane is not necessarily the lane with the lowest
```
```suggestion
- `default_lane` is in `lanes` (the default lane is not necessarily the lane with the lowest
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+priorities and for `CheckTx` responses of invalid transactions.
+
+On receiving the information from the app, CometBFT will validate that:
+- `lanes` has no duplicates (values in `lanes` don't need to be sorted),
```
Would it make our life easier to require that the lanes are sorted (to simplify the validation code)?

hvanz replied:
Not really. It's easy to sort the list and we wouldn't need to trust the app to do it.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+If `lane` is 0, it means that the application did not set any lane in the response message, so the
+transaction will be assigned to the default lane.
+
+### Removing transactions from the mempool
```
Should this also be present in the list at the beginning of the section?

hvanz replied:
I didn't add it because there's nothing important about removing transactions.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+version of the application they are executing. Since the application includes the lane definitions,
+both nodes will also agree on the lanes they implement. Therefore we don't need to modify the P2P
+handshake.
+
```
We could add a paragraph here to explain compatibility with CosmoVisor-like approaches.

hvanz replied:
Yes. That's important and I forgot.

sergio-mena replied:
Will add a paragraph

sergio-mena replied:
Working on it now...

sergio-mena replied:
Done. Please take a look

hvanz replied:
Thanks!

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+
+If the application does not implement lanes (that is, it responds with empty values in
+`InfoResponse`), then `defaultLane` will be set to 1, and `lanes` will have only one entry for the
+default lane.
```
```suggestion
default lane. In this case, the new mempool's behaviour will be equivalent to that of the current mempool.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+- `MaxTxsBytes`, the maximum total number of bytes of the mempool, and
+- `MaxTxBytes`, the maximum size in bytes of a single transaction accepted into the mempool.
+
+With lanes, the total size of the mempool will be the sum of the sizes of all lanes.
```
Do we want to divide the total size evenly across lanes? (or are we gonna expunge a TX from a lower lane when a full mempool receives a TX?)

hvanz replied:
I'd say it's not needed for the MVP but it could be an interesting improvement for the future. Equal sizes for all lanes may not make sense when lanes are defined by transaction size. For example, a lane for only very big txs. Probably we won't see this issue when we add custom configurations per lane.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+With lanes, the total size of the mempool will be the sum of the sizes of all lanes.
+
+Additionally, the `Recheck` and `Broadcast` flags will apply to all lanes or to none. Remember that,
+with `PrepareProposal`, it becomes _always_ mandatory to recheck remaining transactions in the
+mempool, so there is no point in disabling `Recheck` per lane.
```
```suggestion
if `PrepareProposal`'s app logic can ever add a new transaction, it becomes _always_ mandatory to recheck remaining transactions in the
mempool, so there is no point in disabling `Recheck` per lane.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+MVP implementing a variant of the [Weighted Round Robin (WRR)](wrr) algorithm, which meets these
+requirements and is straightforward to implement.
+
+Before modifying the dessemination logic, we need to refactor the current implementation and the
```
```suggestion
Before modifying the dissemination logic, we need to refactor the current implementation and the
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+correct data. A node may take advantage of the network by sending lower-priority transactions before
+higher-priority ones. Although the receiving node could easily verify the priority of a transaction
+when it calls `CheckTx`, it cannot detect if a peer is sending transactions out of order over a
+single P2P channel. For the moment, we leave out of the MVP any mechanism to detect and possible
```
```suggestion
single P2P channel. For the moment, we leave out of the MVP any mechanism to detect and possibly
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+
+The duality lane/priority could introduce a powerful indirection. The app could just define the lane
+of a transaction in `CheckTx`, but the priority of the lane itself could be configured (and
+fine-tuned) elsewhere. For example, by the app itself or by node operators. The proposed design does
```
```suggestion
fine-tuned) elsewhere. For example, by the app itself or by node operators. The proposed design for the MVP does
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+
+We have considered two alternative approaches for _where_ to configure lanes and priorities:
+1. In `config.toml` or `app.toml`. We have discarded this option as it does not make sense for
+   different nodes to have different lane configurations.
```
```suggestion
   different nodes to have different lane configurations. The properties defined in the specification above
   are end-to-end, and so, the lane configuration has to be consistent across the network.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+1. In `ConsensusParams`. There are several disadvantages with this approach. If we allow changing
+   lane information via `ConsensusParams`, the mempool would need to update lanes dynamically. The
+   updating process would be very complex and cumbersome, and not really appealing for an MVP. Two
+   governance proposals would be required to pass to update the lane definitions. A first proposal
```
```suggestion
   governance proposals would be required to pass to update (a) the lane definitions, and (b) a new version of the app with adapted classification logic.
   A first proposal
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+   governance proposals would be required to pass to update the lane definitions. A first proposal
+   would be required for upgrading the application, because the lane classification logic (thus the
+   application's code) needs to know the lane configuration beforehand. And a second proposal would
+   be needed for upgrading the lanes via `ConsensusParams`.
```
```suggestion
   be needed for upgrading the lanes via `ConsensusParams`.
   While it is true that SDK applications could pass a governance proposal with both elements together,
   it would be something to _always_ do, and it is not clear what the situation would be for non-SDK applications.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+  current implementation. 
+- Lanes will preserve the "FIFO ordering of transactions" property within the same class (with a
+  best effort approach, as the current implementation).
+
```
```suggestion
- Applications that are unaware of this feature, and therefore not classifying transactions in `CheckTX`
  will observe the same behavior from the mempool as the current implementation.  
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+- The total FIFO ordering of transaction dissemination and block creation may be broken when
+  multiple lanes are used. Since FIFO ordering is important within the same class of transactions,
+  we expect this will not be a real problem.
+
```
```suggestion
- Increased complexity in the logic of `CheckTx` (ante handlers) in order to classify transactions,
  with a possibility of introducing bugs in the classification logic.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+- Lanes are optional. Current applications do not need to make any change to their code. Future
+  applications will not be forced to use the lanes feature.
+- The proposed prioritization logic (WRR) for the dissemination algorithm is fair, so low-priority
+  transactions will not get stuck in the mempool long periods of time.
```
```suggestion
  transactions will not get stuck in the mempool for long periods of time.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,284 @@ TODO
+_weight_, ensuring each lane gets a fraction of the P2P channel capacity proportional to its
+priority. Moreover, it should be _fair_, to prevent starvation of low-priority lanes. Given the
+extensive research on scheduling algorithms in operating systems and networking, we propose for the
+MVP implementing a variant of the [Weighted Round Robin (WRR)](wrr) algorithm, which meets these
```
In QoS for IP networks there is [WRED](https://en.wikipedia.org/wiki/Weighted_random_early_detection). I need check if that would be relevant here.

cason replied:
The link is broken.

cason replied:
Again, is this what the p2p channel does?

sergio-mena replied:
> The link is broken.

Which link do you refer to? The one on the text, or the one in my comment (both seem to be working for me)

> Again, is this what the p2p channel does?

Yeah, I had the same question when I read this part of Thu. Is priority in p2p channels implementing `wrr` or something else?

hvanz replied:
The link on WRR was broken and then I fixed it. I added a section called Prioritization Logic with a description of the algorithm in P2P channels.

melekes reviewed (approved):
Did a cursory review. Great work 🚀 

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExeWx0aGVjZG90cmdtanIybGRueTc3NDI2bndrZXd1dHRodnlnZGh6eSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/MAzunB1Ru6zAYlYgPD/giphy.gif)

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+impact if the committed block contains few transactions. Therefore, we decided to exclude it from
+the MVP.
+
+### Reaping transactions for creating blocks
```
Any discussion regarding starvation?

Namely, no guarantee that a tx in a low priority lane will eventually be included in the output?

I mentioned that to you guys at some point, and I still think that the way the `p2p/MConnection` does that is probably more adequate. I think that considering this approach should at least be discussed here, as it is an ADR.

hvanz replied:
It was not mentioned so I added it. In essence it's the same as for tx dissemination. I agree we should discuss what P2P does and compare it to the proposed scheduling algorithm.

hvanz replied:
I added a section Prioritization logic that includes the current scheduling algorithm in the P2P reactor.

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+
+### One CList for all lanes
+
+We briefly considered sharing one CList for all lanes, changing the internal logic of CList to
```
This is the priority mempool, in other words.

hvanz replied:
I wouldn't say so because in the priority mempool every tx has a priority. The idea of the "one CList for all lanes" solution was something like adding a lane field to each entry and then other auxiliary data structures to support traversing the entries per lane.

cason replied:
So it is not one clist for all lanes, but one clist and multiple clists, one per lane, using the same entries as the one big clist.

In other words, a linked-list that contains other abstract linked-lists inside it.

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+accommodate lane requirements. However, this design significantly increases code complexity,
+particularly in the transaction dissemination logic.
+
+### One P2P channel per lane
```
Also, we would need a different routine per lane/channel, otherwise there is no real difference between the approaches.

Also consider that the assignment of channels is **static**.

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+channels. Since lanes are defined in the application logic, the nodes only need to agree on the
+application version, as it already happens in the current implementation.
+
+### Duality lane/priority
```
To confirm, now lineID == priority? And the higher the greater?

hvanz replied:
That's right.

cason replied:
I assume this is written somewhere...

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
 
 ## Consequences
 
 TODO
```
Remove?

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+  latency.
+- The mempool will be able to offer Quality of Service guarantees, which does not exist in the
+  current implementation. 
+- Lanes will preserve the "FIFO ordering of transactions" property within the same class (with a
```
This is neutral to me.

hvanz replied:
Changed to neutral.

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+  will observe the same behavior from the mempool as the current implementation.  
 ### Negative
 
+- The total FIFO ordering of transaction dissemination and block creation may be broken when
```
Total FIFO? Eheheh

hvanz replied:
Not what I meant :stuck_out_tongue_closed_eyes:. I rephrased.

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+  best effort approach, as the current implementation).
+- Applications that are unaware of this feature, and therefore not classifying transactions in `CheckTX`
+  will observe the same behavior from the mempool as the current implementation.  
 ### Negative
```
On negative, again, are we discussing the starvation problem of low priority lanes. I refer to the `Reap()` part, not the broadcast part, which seems to be implemented using a starvation-free algorithm.

hvanz replied:
The Reap section did not mention the starvation-free property. I corrected it.

cason commented:
> The first part of this PR is clearly a spec. 

And the use of FIFO order, partial order there is something to review. Lets come up with new names, lets having a subsection about ordering, please, because this is a source of confusion.

hvanz commented on docs/references/architecture/adr-113-mempool-lanes.md:
````diff
@@ -294,254 +296,289 @@ TODO
+```
+The callback that handles the first-time CheckTx response will append the new mempool entry to the
+corresponding `CList`, namely `lanes[lane]`, and update the other auxiliary variables accordingly.
+If `lane` is 0, it means that the application did not set any lane in the response message, so the
````
The application implicitly returns 0 when it ignores lanes in the CheckTx response. We cannot force the application to return a specific value when they don't want to implement lanes, so the mempool needs to take care of assigning the default lane.

hvanz commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+
+### Configuration
+
+For an MVP, all lanes can share the same mempool configuration. In this scenario, all lanes will be
```
You're right. I rephrased it.

hvanz replied:
Agree, this is not clear or correct. I rephrased it.

hvanz commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,289 @@ TODO
+
+### Initialization
+
+Currently, querying the app for `Info` happens during the handshake between CometBFT and the app,
```
Not specifically this scenario but section "Validating lanes of received transactions" is about what to do when a node decides to broadcast txs in whatever order.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,285 @@ TODO
+  priority), and
+- the list `lanes` is empty if and only if `default_value` is 0.
+
+### Initialization
```
Taking care of this...

sergio-mena replied:
Done

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -481,19 +478,42 @@ particular, `TxsFront` is leaking implementation details outside the `Mempool` i
+  `recentlySent/Priority` is the least.
+
+From the extensive research in operating systems and networking, we can choose for the MVP an
+existing scheduling algorithm that meet these requirements and are straightforward to implement,
```
```suggestion
existing scheduling algorithm that meets these requirements and are straightforward to implement,
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -481,19 +478,42 @@ particular, `TxsFront` is leaking implementation details outside the `Mempool` i
+
+From the extensive research in operating systems and networking, we can choose for the MVP an
+existing scheduling algorithm that meet these requirements and are straightforward to implement,
+such as a variant of the [Weighted Round Robin][wrr] (WRR) algorithm. Another one to consider is
+[Weighted Random Early Detection][wred].
```
```suggestion
such as a variant of the [Weighted Round Robin][wrr] (WRR) algorithm.
```

I was wrong w.r.t. WRED, it's rather used to help congestion control algorithms (in the presence of QoS), so not really related here.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -610,7 +630,9 @@ late, possibly having lane definitions that do not match with those of nodes at
 - Skip's [Block SDK][skip-block-sdk]
+- P2P's [selectChannelToGossipOn][selectChannelToGossipOn] function
 - [Weighted Round Robin][wrr]
+- [Weighted Random Early Detection][wred]
```
```suggestion
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -621,5 +643,7 @@ late, possibly having lane definitions that do not match with those of nodes at
 [wrr]: https://en.wikipedia.org/wiki/Weighted_round_robin
-[cosmovisor]: https://docs.cosmos.network/v0.50/build/tooling/cosmovisor
\ No newline at end of file
+[wred]: https://en.wikipedia.org/wiki/Weighted_random_early_detection
```
```suggestion
```

hvanz commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -294,254 +296,298 @@ TODO
+need to set up the lanes when starting up (whether afresh or in recovery mode).
+
+Different nodes also need to agree on the lanes they use. When a node connects to a peer, they both
+perform a handshake to agree on some basic information (see `DefaultNodeInfo`), including the
+version of the application they are executing. Since the application includes the lane definitions,
```
Oops I assumed wrongly. I changed this paragraph proposing to "modify the P2P handshake process to validate that the application versions are compatible". Does this change break something?

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
````diff
@@ -294,254 +296,289 @@ TODO
+include the following fields. These fields need to be filled by the application only in case it
+wants to implement lanes.
+```protobuf
+message InfoResponse {
````
The one from lines 327 to 331.

By the conditions on lines 342 and 343.

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -300,44 +300,49 @@ This section describes the architectural changes needed to implement an MVP of l
 message InfoResponse {
   ...
-  repeated uint32 lanes = 6;
-  uint32 default_lane = 7;
```
Is it correct to set the default lane to 7 if it is not in the list of lanes?

hvanz replied:
`6` and `7` are not default values. They are the field number. Currently `InfoResponse` has 5 fields.

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -300,44 +300,49 @@ This section describes the architectural changes needed to implement an MVP of l
-  priority), and
-- the list `lanes` is empty if and only if `default_value` is 0.
+- `lane_priorities` has no duplicates (values in `lane_priorities` don't need to be sorted),
+- `default_lane_priority` is in `lane_priorities` (the default lane is not necessarily the lane with
```
This is violated in the example above.

hvanz replied:
Same answer as above.

cason commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -352,10 +357,22 @@ for updating the software. CometBFT will not need to deal with dynamic lane chan
+
+Finally, this approach is compatible with applications that need to swap binaries when
+catching up or upgrading, such as SDK applications using [Cosmovisor][cosmovisor].
+When a node is catching up (i.e. BlockSyncing), its peers will detect
```
```suggestion
When a node is catching up (i.e., state or block syncing), its peers will detect
```

cason reviewed (approved):
Reviewing the changes from my previous review, most questions were addressed, great work :)

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+# Mempool cache
+
+The *cache* stores the hashes of recently seen transactions. Its purpose is to avoid processing and
```
Does it only store it nowadays?

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+# Mempool cache
+
+The *cache* stores the hashes of recently seen transactions. Its purpose is to avoid processing and
+adding duplicate transactions to the mempool.
```
Also broadcasting them again.

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+
+## Adding transactions to the cache
+
+A node that receives a transaction, before trying to add it to the mempool, puts it in the cache, if
```
I tend to understand "receives a transaction" via network. Also via RPC, i.e., from clients?

cason replied:
I would rephrase in a simpler way:

1. New transaction, check if it is on the cache
2. If it is, does nothing (what error message is produced?)
3. If it is not, add it to the cache to then start the process of adding it to the mempool.

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+it is not already there. Only if it is not in the cache, the node will attempt to add it to the
+mempool. If the transaction exists in the cache, it is discarded with an error message.
+
+Another way a transaction may enter the cache is when the node's execution module updates the
```
This is another use we have from the cache.

We want to avoid keeping in the mempool transactions that were already committed. Because the mempool is a list of _pending_ or _oustanding_ transactions.

This should be clear also in the introduction.

cason replied:
Again, to rephrase in a simpler way:

1. We want to make sure that we don't process transactions that were already committed
2. So we keep on the cache (hashes) of recently committed transactions
3. Upon the update call, if some committed and valid transaction is not yet present in the cache, we add it to the cache.

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+node receives `tx` and `tx` is not in the cache.
+
+Because of how transactions are evicted from the cache (see below), it is possible by this
+definition that a transaction is received for the first time on multiple occasions.
```
This sentence.... first time multiple times?

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+
+## Capacity
+
+The cache has a *capacity*, a maximum number of transactions that it can hold. The cache's capacity
```
This mixes definitions with recommendations?

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+MUST be bigger than the mempool's capacity (in number of transactions). For example, in the current
+implementation, the cache doubles the capacity of the mempool. 
+
+The cache enforces a FIFO policy for keeping transactions: a transaction will be evicted from the
```
I thought it was a LRU cache. If it is was, there is not FIFO.

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+implementation, the cache doubles the capacity of the mempool. 
+
+The cache enforces a FIFO policy for keeping transactions: a transaction will be evicted from the
+cache if it is the cache's oldest entry and a new transaction is being added. Therefore, a
```
Define oldest, namely, the cache eviction: last added, last accessed, last modified?

cason replied:
This can be helpful: https://en.wikipedia.org/wiki/Cache_replacement_policies

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+
+## Removing transactions from the cache
+
+A transaction is removed from the cache also when the application deems the transaction as invalid.
```
So, the assumption is that the cache keeps _valid_ transactions only? Maybe rephrasing in this way helps the reader?

cason commented on spec/mempool/cache.md:
```diff
@@ -0,0 +1,36 @@
+## Removing transactions from the cache
+
+A transaction is removed from the cache also when the application deems the transaction as invalid.
+Because transactions are evicted when the cache is full, a transaction could be removed from the
```
This has nothing to do with being invalid or not.

Plus, I thought this behavior was configurable, with a config flag.

cason reviewed (commented):
A new spec file popped up, some (several) comments regarding it.

Maybe a different PR would be a better approach.

hvanz commented:
@cason Thanks for all the feedback. I removed the cache spec; it does not belong here and I will leave it for another PR. For now, I simplified the explanation of the "received for the first time" definition, which was the reason to include the cache spec.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -53,32 +55,43 @@ is an orthogonal problem to transaction classification, although the latter may
+:memo: _Definition_: Given any two different transactions `tx1` and `tx2`, in a given node, we say that:
+1. `tx1` is *validated before* `tx2`, when `tx1` and `tx2` are received for the first time, and `tx1`
+is validated against the application (via `CheckTx`) before `tx2`,
+1. `tx1` is *rechecked before* `tx2`, when `tx` and `tx2` are in the mempool and `tx1` is
```
```suggestion
1. `tx1` is *rechecked before* `tx2`, when `tx1` and `tx2` are in the mempool and `tx1` is
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -53,32 +55,43 @@ is an orthogonal problem to transaction classification, although the latter may
+:memo: _Definition_: Given any two different transactions `tx1` and `tx2`, in a given node, we say that:
+1. `tx1` is *validated before* `tx2`, when `tx1` and `tx2` are received for the first time, and `tx1`
+is validated against the application (via `CheckTx`) before `tx2`,
+1. `tx1` is *rechecked before* `tx2`, when `tx` and `tx2` are in the mempool and `tx1` is
+re-validated (rechecked via `CheckTx`) before `tx2`,
```
What if `tx1` and `tx2` are re-checked in different heights? Does that case fit into this definition, or we're leaving such a case out on purpose?

hvanz replied:
If both transactions are in the mempool, that means that they are rechecked at the same height. I'll clarify in the text.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -53,32 +55,43 @@ is an orthogonal problem to transaction classification, although the latter may
+1. `tx1` is *disseminated before* `tx2`, when `tx1` is sent to a given peer before `tx2`.
+
+In 4, note that in the current implementation there is one dissemination routine per peer, so it
+could happen that `tx2` is sent to a peer before `tx1` is sent to a different peer.
```
```suggestion
could happen that `tx2` is sent to a peer before `tx1` is sent to a different peer.
Hence the importance of expression "to a given peer" in that definition.
```

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -53,32 +55,43 @@ is an orthogonal problem to transaction classification, although the latter may
 :parking: _Property_ **FIFO ordering of transactions**: We say that the mempool makes a best effort
-in maintaining the FIFO ordering of transactions when transactions are validated, processed, and
-disseminated in the same order in which the mempool has received them.
+in maintaining the FIFO ordering of transactions when transactions are validated, rechecked, reaped,
```
Excellent!

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -96,27 +109,32 @@ Before stating this property, we need to make some definitions.
 :memo: _Definition_: Each class has a *priority* and two classes cannot have the same priority.
 Therefore all classes can be ordered by priority.
 
+When a transaction is received for the first time and validated via `CheckTx`, the application MUST
```
But, we say somewhere that, if the add doesn't return a class, then the TX belongs to the default class?

hvanz replied:
True. I fixed it.

sergio-mena commented on docs/references/architecture/adr-113-mempool-lanes.md:
```diff
@@ -126,22 +144,22 @@ timestamp of the block in which the transaction finally was included.
 
-This means that some pairs of
-transactions are comparable and, thus, have and order, while others not.
+This means that some pairs of transactions are comparable and, thus, have and order, while others
```
```suggestion
This means that some pairs of transactions are comparable and, thus, have an order, while others
```

sergio-mena reviewed (approved):
🚀
